common/src/main/java/io/pravega/common/concurrent/AsyncSemaphore.java

```diff
@@ -0,0 +1,220 @@
+            }
+        }
+
+        if (toCancel != null) {
+            toCancel.forEach(q -> q.result.cancel(true));
+        }
+    }
+
+    //endregion
+
+    //region Operations
+
+    /**
+     * Executes the given task which requires the given number of credits.
+     *
+     * If there are sufficient credits available for this task to run, it will be invoked synchronously and the returned
+     * result is directly provided by the given task.
+     *
+     * If there are insufficient credits available for this task to run, it will be queued up and executed when credits
+     * become availble. There is no prioritization of queued tasks - they are triggered in the order in which they
+     * are queued up.
+     *
+     * A task will allocate the requested credits when it is triggered. If the task fails (synchronously or asynchronously),
+     * then the requested credits are automatically released back into the pool. If the task succeeds, the credits will
+     * remain.
+     *
+     * @param task    A {@link Supplier} that, when invoked, will execute the task.
+     * @param credits The number of credits this task requires.
+     * @param <T>     Return type.
+     * @return A CompletableFuture that, when completed, will contain the result of the executed task. If the task failed
+     * or was rejected (i.e., due to {@link AsyncSemaphore} closing), it will be failed with the appropriate exception.
+     */
+    public <T> CompletableFuture<T> run(@NonNull Supplier<CompletableFuture<T>> task, int credits) {
+        Preconditions.checkArgument(credits >= 0 && credits <= this.totalCredits,
+                "credits must be a non-negative number smaller than or equal to %s.", this.totalCredits);
+
+        DelayedTask<T> dt;
+        synchronized (this.queue) {
+            Exceptions.checkNotClosed(this.closed, this);
+            if (canExecute(credits)) {
+                dt = null;
+                this.usedCredits += credits;
+            } else {
+                // Insufficient credits; need to queue up and execute when more becomes available.
+                dt = new DelayedTask<>(credits, task);
+                this.queue.addLast(dt);
+            }
+
+        }
+        if (dt == null) {
+            // We have more credits than what this task requires. Execute now without queuing.
+            return execute(task, credits);
+        } else {
+            // This wil be completed when its associated task is executed.
+            return dt.result;
+        }
+    }
+
+    /**
+     * Releases a number of credits back into the pool.
+     *
+     * @param credits The number of credits to release.
+     */
+    public void release(int credits) {
+        Preconditions.checkArgument(credits >= 0, "credits must be a non-negative number.");
+        synchronized (this.queue) {
+            Exceptions.checkNotClosed(this.closed, this);
+            this.usedCredits = Math.max(0, this.usedCredits - credits);
```

Comment:
I updated the Javadoc to explain that the number of credits released will be capped at the number of currently used credits.

I am not going to throw if this is trying to release too much as the caller who releases has no fault that they are giving back more than is used. There is no point in adding an artificial error that will only cause trouble in the future; I consider it better to be flexible.

Reply:
If a bug like your described really exists, then it will immediately become a problem after a failover and recovery. We keep track of this value in parallel (persist it in a segment attribute) and update it every time we have a Table Update and every time we index one or more updates. Upon every failover the `AsyncSemaphore` is reinitialized with the actual value that is stored there. 

This is designed as a way to track available capacity in a system that first queues something, then processes it (and thus releases capacity). You can't process something before you queue it, so really your scenario can't happen.

Reply:
Ok, at least mention the cap here.